### ISSUE TITLE
The init command should default to the global --config when no --config-path is provided

### ISSUE THREAD

**Issue by paltherr** [twpayne/chezmoi#3127]
## Describe the bug

I have two chezmoi repositories under the same account. The first repository uses the default config file in `~/.config/chezmoi/chezmoi.toml`. The second repository uses a custom config file. To run any command for the second repository I have to run chezmoi with the global flag `--config <path-to-custom-config-file>`. As far as I can tell, this works for all commands except for the `init` command. Whenever I modify the `.chezmoi.toml.tmpl` file, chezmoi tells me that I have to rerun the `init` command. However, if I do so without providing the `init` flag `--config-path <path-to-custom-config-file>`, it fails to write to the custom config file and instead writes to `~/.config/chezmoi/chezmoi.toml` and corrupts my first chezmoi repository.

## To reproduce

The attached file contains the full script displayed below.

```console
$ rm -rfI test-chezmoi || exit
$ mkdir test-chezmoi
$ cd test-chezmoi
$ home=$PWD
$ mkdir config
$ git init athome
Initialized empty Git repository in /Users/paltherr/test-chezmoi/athome/.git/
$ echo "sourceDir = \"$home/athome\"" >> athome/.chezmoi.toml.tmpl
$ echo "destDir = \"$home\"" >> athome/.chezmoi.toml.tmpl
$ chezmoi --config $home/config/athome.toml --source $home/athome init --config-path $home/config/athome.toml
$ chezmoi --config $home/config/athome.toml status
$ chezmoi --config $home/config/athome.toml doctor
```

<details>

```console
RESULT    CHECK                MESSAGE
ok        version              v2.36.0, commit be6e90307f9b5c8fa5d356183418a16d7bba69bf, built at 2023-07-28T17:59:29Z, built by Homebrew
ok        latest-version       v2.36.0
ok        os-arch              darwin/amd64
ok        uname                Darwin maloja 22.5.0 Darwin Kernel Version 22.5.0: Mon Apr 24 20:51:50 PDT 2023; root:xnu-8796.121.2~5/RELEASE_X86_64 x86_64
ok        go-version           go1.20.6 (gc)
ok        executable           /usr/local/bin/chezmoi
ok        upgrade-method       replace-executable
failed    config-file          found ~/.config/chezmoi/chezmoi.toml, expected ~/test-chezmoi/config/athome.toml
warning   source-dir           ~/test-chezmoi/athome is a git working tree (dirty)
ok        suspicious-entries   no suspicious entries
warning   working-tree         ~/test-chezmoi/athome is a git working tree (dirty)
ok        dest-dir             ~/test-chezmoi is a directory
ok        umask                022
ok        cd-command           found /bin/zsh
ok        cd-args              /bin/zsh
info      diff-command         not set
ok        edit-command         found /usr/bin/vi
ok        edit-args            /usr/bin/vi
ok        git-command          found /usr/local/bin/git, version 2.41.0
ok        merge-command        found /usr/bin/vimdiff
ok        shell-command        found /bin/zsh
ok        shell-args           /bin/zsh
info      age-command          age not found in $PATH
info      gpg-command          gpg not found in $PATH
info      pinentry-command     not set
info      1password-command    op not found in $PATH
info      bitwarden-command    bw not found in $PATH
info      dashlane-command     dcli not found in $PATH
info      gopass-command       gopass not found in $PATH
info      keepassxc-command    keepassxc-cli not found in $PATH
info      keepassxc-db         not set
info      keeper-command       keeper not found in $PATH
info      lastpass-command     lpass not found in $PATH
info      pass-command         pass not found in $PATH
info      passhole-command     ph not found in $PATH
info      rbw-command          rbw not found in $PATH
info      vault-command        vault not found in $PATH
info      vlt-command          vlt not found in $PATH
info      secret-command       not set
```

</details>

```console
$ ls -la config
total 72
drwxr-xr-x@ 4 paltherr  staff    128 Jul 29 03:46 .
drwxr-xr-x@ 4 paltherr  staff    128 Jul 29 03:46 ..
-rw-------@ 1 paltherr  staff     91 Jul 29 03:46 athome.toml
-rw-------@ 1 paltherr  staff  32768 Jul 29 03:46 chezmoistate.boltdb
$ cat config/athome.toml
sourceDir = "/Users/paltherr/test-chezmoi/athome"
destDir = "/Users/paltherr/test-chezmoi"
```

At this point everything is OK and I can use my second chezmoi repository as expected. I just have to include `--config $home/config/athome.toml` in each command. Interestingly the `doctor` command complains about `config-file` (see details above) but nevertheless everything works as expected.

If I now modify the `.chezmoi.toml.tmpl` file, chezmoi tells me to rerun `init`:

```console
$ echo "# foo" >> athome/.chezmoi.toml.tmpl
$ chezmoi --config $home/config/athome.toml status
chezmoi: warning: config file template has changed, run chezmoi init to regenerate config file
```

If I do so with a `--config-path` then the custom config file gets updated and everything gets back to normal (but `doctor` still complains about `config-file`).

```console
$ chezmoi --config $home/config/athome.toml --source $home/athome init --config-path $home/config/athome.toml
$ chezmoi --config $home/config/athome.toml status
$ chezmoi --config $home/config/athome.toml doctor
```

<details>

```console
RESULT    CHECK                MESSAGE
ok        version              v2.36.0, commit be6e90307f9b5c8fa5d356183418a16d7bba69bf, built at 2023-07-28T17:59:29Z, built by Homebrew
ok        latest-version       v2.36.0
ok        os-arch              darwin/amd64
ok        uname                Darwin maloja 22.5.0 Darwin Kernel Version 22.5.0: Mon Apr 24 20:51:50 PDT 2023; root:xnu-8796.121.2~5/RELEASE_X86_64 x86_64
ok        go-version           go1.20.6 (gc)
ok        executable           /usr/local/bin/chezmoi
ok        upgrade-method       replace-executable
failed    config-file          found ~/.config/chezmoi/chezmoi.toml, expected ~/test-chezmoi/config/athome.toml
warning   source-dir           ~/test-chezmoi/athome is a git working tree (dirty)
ok        suspicious-entries   no suspicious entries
warning   working-tree         ~/test-chezmoi/athome is a git working tree (dirty)
ok        dest-dir             ~/test-chezmoi is a directory
ok        umask                022
ok        cd-command           found /bin/zsh
ok        cd-args              /bin/zsh
info      diff-command         not set
ok        edit-command         found /usr/bin/vi
ok        edit-args            /usr/bin/vi
ok        git-command          found /usr/local/bin/git, version 2.41.0
ok        merge-command        found /usr/bin/vimdiff
ok        shell-command        found /bin/zsh
ok        shell-args           /bin/zsh
info      age-command          age not found in $PATH
info      gpg-command          gpg not found in $PATH
info      pinentry-command     not set
info      1password-command    op not found in $PATH
info      bitwarden-command    bw not found in $PATH
info      dashlane-command     dcli not found in $PATH
info      gopass-command       gopass not found in $PATH
info      keepassxc-command    keepassxc-cli not found in $PATH
info      keepassxc-db         not set
info      keeper-command       keeper not found in $PATH
info      lastpass-command     lpass not found in $PATH
info      pass-command         pass not found in $PATH
info      passhole-command     ph not found in $PATH
info      rbw-command          rbw not found in $PATH
info      vault-command        vault not found in $PATH
info      vlt-command          vlt not found in $PATH
info      secret-command       not set
```
</details>

```console
$ ls -la config
total 72
drwxr-xr-x@ 4 paltherr  staff    128 Jul 29 03:46 .
drwxr-xr-x@ 4 paltherr  staff    128 Jul 29 03:46 ..
-rw-------@ 1 paltherr  staff     97 Jul 29 03:46 athome.toml
-rw-------@ 1 paltherr  staff  32768 Jul 29 03:46 chezmoistate.boltdb
$ cat config/athome.toml
sourceDir = "/Users/paltherr/test-chezmoi/athome"
destDir = "/Users/paltherr/test-chezmoi"
# foo
```

If I modify again the `.chezmoi.toml.tmpl` file but rerun `init` with no `--config-path` then chezmoi fails to update the custom config file. Interestingly it still updates the `chezmoistate.boltdb` file next to custom config file (which can't be seen in the output here but I could observe that its timestamp was updated when I ran the command manually). It's also interesting to note that afterwards the `status` command no longer asks to rerun `init`.

```console
$ echo "# bar" >> athome/.chezmoi.toml.tmpl
$ chezmoi --config $home/config/athome.toml status
chezmoi: warning: config file template has changed, run chezmoi init to regenerate config file
$ chezmoi --config $home/config/athome.toml --source $home/athome init
$ chezmoi --config $home/config/athome.toml status
$ chezmoi --config $home/config/athome.toml doctor
```

<details>

```console
RESULT    CHECK                MESSAGE
ok        version              v2.36.0, commit be6e90307f9b5c8fa5d356183418a16d7bba69bf, built at 2023-07-28T17:59:29Z, built by Homebrew
ok        latest-version       v2.36.0
ok        os-arch              darwin/amd64
ok        uname                Darwin maloja 22.5.0 Darwin Kernel Version 22.5.0: Mon Apr 24 20:51:50 PDT 2023; root:xnu-8796.121.2~5/RELEASE_X86_64 x86_64
ok        go-version           go1.20.6 (gc)
ok        executable           /usr/local/bin/chezmoi
ok        upgrade-method       replace-executable
failed    config-file          found ~/.config/chezmoi/chezmoi.toml, expected ~/test-chezmoi/config/athome.toml
warning   source-dir           ~/test-chezmoi/athome is a git working tree (dirty)
ok        suspicious-entries   no suspicious entries
warning   working-tree         ~/test-chezmoi/athome is a git working tree (dirty)
ok        dest-dir             ~/test-chezmoi is a directory
ok        umask                022
ok        cd-command           found /bin/zsh
ok        cd-args              /bin/zsh
info      diff-command         not set
ok        edit-command         found /usr/bin/vi
ok        edit-args            /usr/bin/vi
ok        git-command          found /usr/local/bin/git, version 2.41.0
ok        merge-command        found /usr/bin/vimdiff
ok        shell-command        found /bin/zsh
ok        shell-args           /bin/zsh
info      age-command          age not found in $PATH
info      gpg-command          gpg not found in $PATH
info      pinentry-command     not set
info      1password-command    op not found in $PATH
info      bitwarden-command    bw not found in $PATH
info      dashlane-command     dcli not found in $PATH
info      gopass-command       gopass not found in $PATH
info      keepassxc-command    keepassxc-cli not found in $PATH
info      keepassxc-db         not set
info      keeper-command       keeper not found in $PATH
info      lastpass-command     lpass not found in $PATH
info      pass-command         pass not found in $PATH
info      passhole-command     ph not found in $PATH
info      rbw-command          rbw not found in $PATH
info      vault-command        vault not found in $PATH
info      vlt-command          vlt not found in $PATH
info      secret-command       not set
```

</details>

```console
$ ls -la config
total 72
drwxr-xr-x@ 4 paltherr  staff    128 Jul 29 03:46 .
drwxr-xr-x@ 4 paltherr  staff    128 Jul 29 03:46 ..
-rw-------@ 1 paltherr  staff     97 Jul 29 03:46 athome.toml
-rw-------@ 1 paltherr  staff  32768 Jul 29 03:46 chezmoistate.boltdb
$ cat config/athome.toml
sourceDir = "/Users/paltherr/test-chezmoi/athome"
destDir = "/Users/paltherr/test-chezmoi"
# foo
```

Instead of updating the custom config file, the `init` command updated the default config file `~/.config/chezmoi/chezmoi.toml`:

```console
$ cat $HOME/.config/chezmoi/chezmoi.toml
sourceDir = "/Users/paltherr/test-chezmoi/athome"
destDir = "/Users/paltherr/test-chezmoi"
# foo
# bar
```

## Expected behavior

The `init` command should default `--config-path` to `--config` if the latter was provided but not the former.

[test-chezmoi.txt](https://github.com/twpayne/chezmoi/files/12206108/test-chezmoi.txt)

**Comment by 2mc** [comment]
CHEZMOI is an exceptional config manager!
It's really 'for me'!

However, I am a split person and really need to be able to have 2 or more separate chezmoi repositories.

Now, the one repository corrupts the other, exactly as it was described above!

Unfortunately, this is a true show stopper.

I could provide more details to help solving this bug.
But the above report is already pretty exhaustive and to the point.

Any comment from the developers on how high or low they consider this bug to be on their priority list, and whether a fix can be expected soon, later, or whenever would be appreciated. 

Thank you!

**Comment by paltherr** [comment]
Are there other commands than `init` that lead to corruption? As far as I know, only `init` is problematic and there is a workaround: you "just" have to remember to add `--config-path <path-to-the-alternate-config>` whenever you run the `init` command.

To make sure that I don't forget it, I wrote a script to use my alternate config. It simply invokes `chezmoi` with its arguments but systematically adds the `chezmoi` option `--config <path-to-the-alternate-config>` and additionally adds the `init` option `--config-path <path-to-the-alternate-config>` in case the `init` command is invoked.

So far, I have never noticed a problem with that setup. However, I would still welcome a fix that removes the need for the extra `--config <path-to-the-alternate-config>`.

**Comment by twpayne** [comment]
Thank you for the excellent bug repot @paltherr and apologies for the slow response on this one. Thank you @2mc for the kind words :)

The tl;dr is that I would like to fix this, but it's not trivial to do so, and as @paltherr has the workaround of using `--config <path-to-the-alternate-config>`, I have not prioritized this.

Here are some of the problems.

1. chezmoi has three sources of configuration information: its builtin defaults, its config file, and its command line flags. Normally they are read in this order, with the command line flags taking priority over any values in its config file. But what should happen if you specify the config file in a command line flag? You've already read the config file. If you read the config file after parsing the command line flags then values in the config file overwrite the command line flags, which is not what you want.

2. chezmoi has multiple potential configuration files. It supports the [XDG specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html), which allows a potentially huge number of user-specified configuration directories. This number of configuration directories is multiplied by all the different config file formats (JSON, TOML, YAML) that chezmoi supports. Just [finding the default config file](https://github.com/twpayne/chezmoi/blob/7c117e72e5bc8b679d639707e217e6d3061ae190/internal/cmd/config.go#L820-L873) is over 50 lines of code. But what happens if the user uses the `--config` flag to specify a config file that happens to be an existing config file? How do you tell if the config file was specified by the user or detected automatically?

I think what is needed is for chezmoi to track the provenance of each configuration value. Right now, chezmoi uses a couple of third-party libraries for parsing command line flags and reading config files, which do not track such provenance. Fundamentally, chezmoi needs to be able to differentiate between "this config value is its default value which was not overridden by the user" and "this config value was set by the user to be exactly its default value".

On top of this, chezmoi needs to change its priority order of command line flags overriding config file values overriding defaults to special case handling of the `--config` flag. Specifically, chezmoi first needs to read its defaults, then check if `--config` is specified and if so change the config file that it reads (even if the `--config` flag specifies a config file that is one of chezmoi's defaults), and only after that overwrite any other config values with whatever is specified in the other command line flags.

Apologies if this does not make much sense. I hope that it communicates just how complex this issue is.

Fundamentally, if chezmoi's configuration values depend on its config file, and chezmoi's config file depends on its command line flags, and chezmoi's command line flags need to override the values from chezmoi's configuration file, what do you do first?
